packages/labs/analyzer/src/lib/lit-html/template.ts

```diff
@@ -163,3 +179,201 @@ const isResolvedIdentifierLitHtmlTemplate = (
   }
   return isKnownLitModuleSpecifier(specifier.text);
 };
+
+export const PartType = {
+  ATTRIBUTE: 1,
+  CHILD: 2,
+  PROPERTY: 3,
+  BOOLEAN_ATTRIBUTE: 4,
+  EVENT: 5,
+  ELEMENT: 6,
+} as const;
+
+export type PartType = (typeof PartType)[keyof typeof PartType];
+
+export interface PartInfo {
+  type: PartType;
+  valueIndex: number;
+}
+
+export interface SinglePartInfo extends PartInfo {
+  type: typeof PartType.CHILD | typeof PartType.ELEMENT;
+  expression: ts.Expression;
+}
+
+export interface AttributePartInfo extends PartInfo {
+  type:
+    | typeof PartType.ATTRIBUTE
+    | typeof PartType.BOOLEAN_ATTRIBUTE
+    | typeof PartType.PROPERTY
+    | typeof PartType.EVENT;
+  prefix: string | undefined;
+  name: string | undefined;
+  strings: string[];
+  expressions: Array<ts.Expression>;
+}
+
+export const hasChildPart = (
+  node: CommentNode
+): node is LitTemplateCommentNode => {
+  return (node as LitTemplateCommentNode).litPart?.type === PartType.CHILD;
+};
+
+export const hasAttributePart = (
+  node: Attribute
+): node is LitTemplateAttribute => {
+  return (node as LitTemplateAttribute).litPart !== undefined;
+};
+
+export type LitTemplateNode = Node & {
+  litNodeIndex: number;
+};
+
+export interface LitTemplate extends DocumentFragment {
+  strings: TemplateStringsArray;
+  parts: Array<PartInfo>;
+}
+
+export interface LitTemplateCommentNode extends CommentNode {
+  litPart?: SinglePartInfo;
+  litNodeIndex: number;
+}
+
+export interface LitTemplateElement extends Element {
+  litNodeIndex: number;
+}
+
+export interface LitTemplateAttribute extends Attribute {
+  litPart: PartInfo;
+}
+
+export const parseLitTemplate = (
+  node: ts.TaggedTemplateExpression,
+  ts: TypeScript,
+  _checker: ts.TypeChecker
+): LitTemplate => {
+  const strings = getTemplateStrings(node, ts);
+  const values = ts.isNoSubstitutionTemplateLiteral(node.template)
+    ? []
+    : node.template.templateSpans.map((s) => s.expression);
+  const parts: Array<PartInfo> = [];
```

Comment:
It might be good to clarify the use cases for the API's here. i.e. whether we intend the DOM AST that comes out to be mutable or not, since the `parts` list will only be the "as-parsed" values.

This could either be made a getter that reifies the parts list each access (slower), or could add an "invalidateParts" API?

And are you planning on adding something to go from a `LitTemplate` back to a `ts.TaggedTemplateExpression`? (that'll presumably need ^^)

Reply:
It's obv fine for some use cases, where you just want to statically analyze only, but for others the `parts` sort of seems not that useful.

Reply:
fwiw in the lint side of things, we don't need to mutate nodes but we do need a way to translate between nodes or positions. i.e. we need a way to go for a parse5 node to an estree node (whether by reference or just by knowing the position)